File: src/minimal_example/src/publisher.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>
#include "minimal_example/GetCounter.h"

void publishHello(ros::Publisher &pub) {
    std_msgs::String msg;
    msg.data = "hello world";
    pub.publish(msg);
}

void getCounter(ros::ServiceClient &client) {
    minimal_example::GetCounter srv;
    if (client.call(srv)) {
        ROS_INFO("Counter: %d", (int) srv.response.counter);
    } else {
        ROS_ERROR("Failed to call service counter");
    }
}

int main(int argc, char **argv) {
    if (argc < 2) { return 1; }
    ros::init(argc, argv, "talker");
    ros::NodeHandle n;
    ros::Publisher pub = n.advertise<std_msgs::String>("monologue", 10);
    ros::ServiceClient client = n.serviceClient<minimal_example::GetCounter>("counter");
    double frequency = -1;
    if (n.getParam(argv[1], frequency)) {
        ROS_INFO("Got frequency from parameter.");
    } else {
        ROS_ERROR("Could not get frequency from parameter.");
        frequency = 10;
    }
    ros::Rate loop_rate(frequency);
    while (ros::ok()) {
        publishHello(pub);
        getCounter(client);
        ros::spinOnce();
        loop_rate.sleep();
    }
    return 0;
}

File: src/minimal_python/src/minimal_python/publisher.py
# -*- coding: utf-8 -*-

# SPDX-License-Identifier: MIT
# Copyright © 2021 André Santos

import rospy
from std_msgs.msg import String
from minimal_example.srv import GetCounter

def publish_hello(pub):
    msg = String()
    msg.data = 'hello world'
    pub.publish(msg)

def get_counter(client):
    try:
        response = client()
        rospy.loginfo('Counter: {}'.format(response.counter))
    except rospy.ServiceException as e:
        rospy.logerr('Failed to call service counter')
        rospy.logerr(repr(e))

def run_node(freq_param):
    rospy.init_node('talker')
    pub = rospy.Publisher('monologue', String, queue_size=10)
    client = rospy.ServiceProxy('counter', GetCounter)
    frequency = rospy.get_param(freq_param, -1)
    if frequency == -1:
        rospy.logerr('Could not get frequency from parameter.')
        frequency = 10
    else:
        rospy.loginfo('Got frequency from parameter.')
    loop_rate = rospy.Rate(frequency)
    while not rospy.is_shutdown():
        publish_hello(pub)
        get_counter(client)
        loop_rate.sleep()

def main(args):
    if len(args) < 1:
        return 1
    try:
        run_node(args[0])
    except rospy.ROSInterruptException:
        pass
    return 0
    

